### PR TITLE
Improve error output from json_is and json_like

### PR DESCRIPTION
**Pull request by DrHyde**
### Summary
Tweaked `Test::Mojo::json_is` and `Test::Mojo::json_like` to detect when a pointer is invalid, and to then just `fail()` instead of comparing data using `is_deeply()` or `like()`. The detection is done using what used to be the core of `json_has` and `json_hasnt`, which has been factored out.

Also added tests for `json_is` and `json_like`.

### Motivation
`...->json($p)` returns `undef` when $p points to lala-land, which leads to `is_deeply()` and `like()` trying to compare the desired value to undef, failing (usually) as expected, *but then erroneously talking about a comparison with undef in their explanations* as if the pointer was to an undef value instead of pointing to nothing at all. This may also lead to confusion when there are legitimate JSON `null`s in the data although I've not tested that.

### References
[issue 1456](https://github.com/mojolicious/mojo/issues/1456)

### CODE REVIEW

File: lib/Test/Mojo.pm [hunk 1]
```diff
@@ -191,31 +191,40 @@ sub header_unlike {
     $regex, $desc);
 }
 
+sub _does_json_contain {
+  my ($self, $p) = @_;
+  Mojo::JSON::Pointer->new($self->tx->res->json)->contains($p);
```
Comment (jhthorsen) [inline lib/Test/Mojo.pm]: This method is not in the right place in the file. All methods are ordered alphabetically. I also think I would rename it so it takes less time to parse what the method does:

```perl
sub _json_contains {
  return Mojo::JSON::Pointer->new(shift->tx->res->json)->contains(shift);
}
```

File: lib/Test/Mojo.pm [hunk 2]
```diff
@@ -191,31 +191,40 @@ sub header_unlike {
...
+}
+
 sub json_has {
   my ($self, $p, $desc) = @_;
   $desc = _desc($desc, qq{has value for JSON Pointer "$p"});
-  return $self->_test('ok',
-    !!Mojo::JSON::Pointer->new($self->tx->res->json)->contains($p), $desc);
+  return $self->_test('ok', !!_does_json_contain($self, $p), $desc);
```
Comment (jhthorsen) [inline lib/Test/Mojo.pm]: Why don't you call a method, instead of a function?

File: lib/Test/Mojo.pm [hunk 3]
```diff
@@ -191,31 +191,40 @@ sub header_unlike {
...
-  return $self->_test('ok',
-    !Mojo::JSON::Pointer->new($self->tx->res->json)->contains($p), $desc);
+  return $self->_test('ok', !_does_json_contain($self, $p), $desc);
 }
 
 sub json_is {
   my $self = shift;
   my ($p, $data) = @_ > 1 ? (shift, shift) : ('', shift);
   my $desc = _desc(shift, qq{exact match for JSON Pointer "$p"});
+  if($p && !_does_json_contain($self, $p)) {
```
Comment (jhthorsen) [inline lib/Test/Mojo.pm]: This PR must be run through perltidy. Pretty sure this if can be condensed into fewer lines of code as well.

File: t/test/mojo.t [hunk 4]
```diff
@@ -0,0 +1,118 @@
+use Mojo::Base -strict;
```
Comment (jhthorsen) [inline t/test/mojo.t]: I'm not sure if this file is required. What does it really test? I can't see that we test Test::Mojo directly anywhere.

**Comment by kraih** [comment]
Afraid i have to disagree with @jhthorsen here, i don't think the little gain is worth maintaining those messy tests.

**Comment by kraih** [comment]
That said, better test coverage for our `Test::Mojo` methods is not a bad idea at all. But if we go there it needs to be consistent.

**Comment by kraih** [comment]
Having thought about this a bit more, i think we need real tests for all `Test::Mojo` methods with their current behaviour before we can think about handling more edge cases. So this PR is blocked until that happened.

**Comment by jhthorsen** [comment]
Seems like this PR has gone stale.

I also agree that figuring out #1460 first is a good idea.

**Comment by DrHyde** [comment]
On the whole I agree that having comprehensive tests before applying this would be a good thing, I just (currently) lack the time to create them. It's on my rather long to-do list.
